core/arch/riscv64/mangle.c

```diff
@@ -619,22 +619,29 @@ mangle_rel_addr(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr,
     instr_get_rel_data_or_instr_target(instr, &tgt);
     ASSERT(opnd_is_reg(dst));
     ASSERT(opnd_is_rel_addr(instr_get_src(instr, 0)));
+    bool uses_reg_stolen = instr_uses_reg(instr, dr_reg_stolen),
+         uses_reg_tp = instr_uses_reg(instr, DR_REG_TP);
 
-    ASSERT_NOT_IMPLEMENTED(!instr_uses_reg(instr, DR_REG_TP));
-
-    if (instr_uses_reg(instr, dr_reg_stolen)) {
+    if (uses_reg_stolen || uses_reg_tp) {
         dst = opnd_create_reg(DR_REG_A0);
         PRE(ilist, next_instr,
             instr_create_save_to_tls(dcontext, DR_REG_A0, TLS_REG0_SLOT));
     }
 
     insert_mov_immed_ptrsz(dcontext, (ptr_int_t)tgt, dst, ilist, next_instr, NULL, NULL);
 
-    if (instr_uses_reg(instr, dr_reg_stolen)) {
+    if (uses_reg_stolen) {
         PRE(ilist, next_instr,
             instr_create_save_to_tls(dcontext, DR_REG_A0, TLS_REG_STOLEN_SLOT));
         PRE(ilist, next_instr,
             instr_create_restore_from_tls(dcontext, DR_REG_A0, TLS_REG0_SLOT));
+    } else if (uses_reg_tp) {
+        PRE(ilist, next_instr,
+            instr_create_save_to_tls(
+                dcontext, DR_REG_A0,
+                os_tls_offset(os_get_app_tls_base_offset(TLS_REG_LIB))));
```

Comment:
`os_tls_offset()` should be unnecessary for RISC-V-specific code, so I think it's a bit cleaner not to use it. (There are some uses of this function in #7062 too, please change that too if it makes sense.)

Reply:
> `os_tls_offset()` should be unnecessary for RISC-V-specific code, so I think it's a bit cleaner not to use it. (There are some uses of this function in #7062 too, please change that too if it makes sense.)

Though `os_tls_offset()` is just nop on RISC-V, I think keeping its usage sync with other architectures helps understand how the code works. Dropping it does save a few bytes but brings assumption of another module, which seems not a good practice.

Reply:
> I think keeping its usage sync with other architectures helps understand how the code works.

Well, it’s not always used in other arch-specific code too, so there is no synchronization to keep, also os_get_app_tls_base_offset() was already used several times in RISC-V-specific code without os_tls_offset().

> Dropping it does save a few bytes but brings assumption of another module, which seems not a good practice.

Well, it’s arch-specific code, it is where you make all kinds of assumptions.

Reply:
It makes sense to me.

In the new commit, `mangle_special_registers()` is reused for context saving, I think this could be marked as resolved.